lib/tasks/book2.rake

```diff
@@ -74,140 +74,147 @@ task :remote_genbook2 => :environment do
-          xlink.each do |link|
-            xref = link.first
-            html.gsub!(/href=\"1-.*?\.html\##{xref}\"/, "href=\"ch00/#{xref}\"") rescue nil
+          if xlink = html.scan(/href=\"1-.*?\.html\#(.*?)\"/)
+            xlink.each do |link|
+              xref = link.first
+              html.gsub!(/href=\"1-.*?\.html\##{xref}\"/, "href=\"ch00/#{xref}\"") rescue nil
+            end
           end
-        end
 
-        if xlink = html.scan(/href=\"\#(.*?)\"/)
-          xlink.each do |link|
-            xref = link.first
-            html.gsub!(/href=\"\##{xref}\"/, "href=\"ch00/#{xref}\"") rescue nil
+          if xlink = html.scan(/href=\"\#(.*?)\"/)
+            xlink.each do |link|
+              xref = link.first
+              html.gsub!(/href=\"\##{xref}\"/, "href=\"ch00/#{xref}\"") rescue nil
+            end
           end
-        end
 
-        if subsec = html.scan(/<img src="(.*?)"/)
-          subsec.each do |sub|
-            sub = sub.first
-            html.gsub!(/<img src="#{sub}"/, "<img src=\"/book/en/v2/#{sub}\"") rescue nil
+          if subsec = html.scan(/<img src="(.*?)"/)
+            subsec.each do |sub|
+              sub = sub.first
+              html.gsub!(/<img src="#{sub}"/, "<img src=\"/book/en/v2/#{sub}\"") rescue nil
+            end
           end
-        end
-
-        puts "\t\t#{chapter_type} #{chapter_number}.#{section} : #{chapter_title} . #{section_title} - #{html.size}"
 
-        csection = schapter.sections.where(:number => section).first_or_create
-        csection.title = section_title.to_s
-        csection.html = pretext + html
-        csection.save
+          puts "\t\t#{chapter_type} #{chapter_number}.#{section} : #{chapter_title} . #{section_title} - #{html.size}"
 
-        xref = Xref.where(:book_id => book.id, :name => id_xref).first_or_create
-        xref.section = csection
-        xref.save
+          csection = schapter.sections.where(:number => section).first_or_create
+          csection.title = section_title.to_s
+          csection.html = pretext + html
+          csection.save
 
-        # record all the xrefs
-        (sec.search(".//*[@id]")).each do |id|
-          id_xref = id.attribute('id').to_s
           xref = Xref.where(:book_id => book.id, :name => id_xref).first_or_create
           xref.section = csection
           xref.save
+
+          # record all the xrefs
+          (sec.search(".//*[@id]")).each do |id|
+            id_xref = id.attribute('id').to_s
+            xref = Xref.where(:book_id => book.id, :name => id_xref).first_or_create
+            xref.section = csection
+            xref.save
+          end
+
+          section += 1
+          pretext = ""
         end
+      end
 
-        section += 1
-        pretext = ""
+      book.sections.each do |section|
+        section.set_slug
+        section.save
       end
+
+    rescue Exception => msg
+      puts msg
     end
+    # consider the revision as processed whatever the result
```

Comment:
The overall goal looks reasonable here, but I am not sure I agree with this "whatever the result". If we get an exception, then either:

  1. It was something temporary and unrelated to the content, and we'd prefer to retry tomorrow.

  2. There's something wrong with the content, but we can work around it by changing the build process. When we do so, we'd want to retry the build.

  3. There's something wrong with the content, which will remain broken until the translation fixes it.

Calling `book.save` on the possibly-broken content hurts the first two. It does save us some effort on the third case (because we would just retry every night), but I don't know how big a deal that is. It is probably a bigger problem that if there is such an exception, nobody would ever be notified. The book content would just silently fail to update.